Update dispute-coordinator documentation (#4975)
paritytech/polkadot: roadmap/implementers-guide/src/node/disputes/dispute-coordinator.md

# Dispute Coordinator

> This is the central subsystem of the node-side components which participate in disputes. This
> subsystem wraps a database which tracks all statements observed by all validators over some window
> of sessions. Votes older than this session window are pruned.

Reviewer: ```suggestion
The coordinator is the central subsystem of the node-side components which participate in disputes. It
wraps a database, which used to track all statements observed by _all_ validators over some window
of sessions. Votes older than this session window are pruned.
```

This subsystem will be the point which produce dispute votes, either positive or negative, based on
locally-observed validation results as well as a sink for votes received by other subsystems. When

> importing a dispute vote from another node, this will trigger participation in the dispute.

Reviewer 2: Better: When a statement import makes it clear that a dispute has been raised (there are opposing votes for a candidate), the dispute coordinator will make sure the local node will participate in the dispute.

This draws on the [dispute statement types][DisputeTypes]

```rust
/// Tracked votes on candidates, for the purposes of dispute resolution.

> #[derive(Debug, Clone, Encode, Decode)]

Reviewer: We usually omit the rust specific annotations.

pub struct CandidateVotes {
	/// The receipt of the candidate itself.
	pub candidate_receipt: CandidateReceipt,
	/// Votes of validity, sorted by validator index.
	pub valid: Vec<(ValidDisputeStatementKind, ValidatorIndex, ValidatorSignature)>,
	/// Votes of invalidity, sorted by validator index.
	pub invalid: Vec<(InvalidDisputeStatementKind, ValidatorIndex, ValidatorSignature)>,
}

/// The mapping for recent disputes; any which have not yet been pruned for being ancient.
pub type RecentDisputes = std::collections::BTreeMap<(SessionIndex, CandidateHash), DisputeStatus>;

/// The status of dispute. This is a state machine which can be altered by the
/// helper methods.
#[derive(Debug, Clone, Copy, Encode, Decode, PartialEq)]
pub enum DisputeStatus {
	/// The dispute is active and unconcluded.
	#[codec(index = 0)]
	Active,

> 	/// The dispute has been concluded in favor of the candidate

Reviewer: ```suggestion
	/// The dispute was concluded in favor of the candidate
```

Author: This comment is copy-pasted from the code. I'd better remove it altogether because it doesn't contribute to understanding how the subsystem works.

	/// since the given timestamp.
	#[codec(index = 1)]
	ConcludedFor(Timestamp),
	/// The dispute has been concluded against the candidate
	/// since the given timestamp.
	///
	/// This takes precedence over `ConcludedFor` in the case that
	/// both are true, which is impossible unless a large amount of
	/// validators are participating on both sides.
	#[codec(index = 2)]
	ConcludedAgainst(Timestamp),

> 	/// Dispute has been confirmed (more than `byzantine_threshold` have already participated/ or

Reviewer: `byzantine_threshold` of _all_ validators.

	relay_parent_block_number: BlockNumber,
	/// By adding the `CandidateHash`, we can guarantee a unique ordering across candidates.
	candidate_hash: CandidateHash,
}
```

It also implements `PartialEq`, `Eq`, `PartialOrd` and `Ord` traits enabling comparison operations
with the comparators.

`Comparator` is used inside `Participation` module as a key for saving `ParticipationRequest`. It
provides the ordering required to process the most important requests first (check the next section
for details).

### Participation
This module keeps track of the disputes that the node participates in. At most there are
`MAX_PARALLEL_PARTICIPATIONS` parallel participations in the subsystem. The internal state of the
module is:

```rust
pub struct Participation {
	/// Participations currently being processed.
	running_participations: HashSet<CandidateHash>,
	/// Priority and best effort queues.
	queue: Queues,
	/// Sender to be passed to worker tasks.
	worker_sender: WorkerMessageSender,
	/// Some recent block for retrieving validation code from chain.
	recent_block: Option<(BlockNumber, Hash)>,
}
```
New candidate is added for processing by calling `fn queue_participation()`. The function either

> starts processing immediately or queues the request. `Participation` uses another internal module

Author: I've reworded it to explain the idea without too many implementation details.

`Queues` which provides prioritisation of the disputes. It guarantees that important disputes will
be processed first. The actual decision how important is a given dispute is performed by the
`ordering` module.

The actual participation is performed by `fn participate()`. First it sends
`AvailabilityRecoveryMessage::RecoverAvailableData` to obtain data from the validators. Then gets
the validation code and stores `AvailableData` with `AvailabilityStoreMessage::StoreAvailableData`
message.  Finally Participation module performs the actual validation and sends the result as
`WorkerMessage` to the main subsystem (`DisputeCoordinatorSubsystem`). `Participation` generates
messages which `DisputeCoordinatorSubsystem` consumes. You can find more information how these
events are processed in the next section.

### SpamSlots

`struct SpamSlots` aims to protect the validator from malicious peers generating erroneous disputes
with the purpose of overloading the validator with unnecessary work.

How the spam protection works? Each peer validator has got a spam slot for unconfirmed disputes with
fixed size (`MAX_SPAM_VOTES`). Each unconfirmed dispute is added to one such slot until all slots

> for the given validator are filled up. At this point all new unconfirmed disputes are ignored.

Reviewer 2: statements from this validator for unconfirmed disputes are ignored .... would be a bit more precise.

What unconfirmed dispute means? Quote from the source code provides an excellent explanation:

> Node has not seen the candidate be included on any chain, it has not cast a
> vote itself on that dispute, the dispute has not yet reached more than a third of
> validator's votes and the including relay chain block has not yet been finalized.

`SpamSlots` has got this internal state:

```rust
pub struct SpamSlots {
	/// Counts per validator and session.
	///
	/// Must not exceed `MAX_SPAM_VOTES`.
	slots: HashMap<(SessionIndex, ValidatorIndex), SpamCount>,

	/// All unconfirmed candidates we are aware of right now.
	unconfirmed: UnconfirmedDisputes,
}
```

It's worth noting that `SpamSlots` provides an interface for adding entries (`fn add_unconfirmed()`)
and removing them (`fn clear()`). The actual spam protection logic resides in the main subsystem, in
`fn handle_import_statements()`. It is invoked during `DisputeCoordinatorMessage::ImportStatements`
message handling (there is a dedicated section for it below). Spam slots are indexed by session id
and validator index. For each such pair there is a limit of active disputes. If this limit is
reached - the import is ignored.

> Spam protection is performed only on invalid vote statements which are not included in any chain,

Reviewer 2: where the concerned candidate is not included on any chain

```rust
struct State {
  keystore: Arc<LocalKeystore>,
	rolling_session_window: RollingSessionWindow,
	highest_session: SessionIndex,
	spam_slots: SpamSlots,
	participation: Participation,
	ordering_provider: OrderingProvider,
	participation_receiver: WorkerMessageReceiver,
	metrics: Metrics,
	// This tracks only rolling session window failures.
	// It can be a `Vec` if the need to track more arises.
	error: Option<SessionsUnavailable>,
	/// Latest relay blocks that have been successfully scraped.
	last_scraped_blocks: LruCache<Hash, ()>,
}
```

### On startup

> Wait for first leaf from Runtime.

Reviewer 2: Leaves don't come from runtime, but the overseer - which gets them on chain import - no runtime calls are involved here.

## Functionality

This assumes a constant `DISPUTE_WINDOW: SessionWindowSize`. This should correspond to at least 1
day.

Ephemeral in-memory state:

```rust
struct State {
  keystore: Arc<LocalKeystore>,

> 	rolling_session_window: RollingSessionWindow,

Reviewer 3: nit: mixture of spaces and tabs
AFAIK, we're using spaces in the .md documentation elsewhere.
